src/queryBuilder/QueryBuilder.js

```diff
@@ -0,0 +1,138 @@
+import React from 'react'
+import * as F from 'futil-js'
+import _ from 'lodash/fp'
+import { observable } from 'mobx'
+import { Provider } from 'mobx-react'
+import DDContext from './DragDrop/DDContext'
+import { Component } from '../utils/mobx-react-utils'
+import Group from './Group'
+import styles from '../styles'
+import { oppositeJoin } from '../utils/search'
+import treeUtils from '../utils/tree'
+let { background } = styles
+let randomString = () =>
+  Math.random()
+    .toString(36)
+    .substring(7)
+
+let { encode } = F.encoder('/') // todo get from client
+
+let blankNode = () => ({ key: randomString() })
+let replaceOn = (list, from, to) => list.splice(list.indexOf(from), 1, to)
+
+let ContextureClientBridge = (Types, Tree) => {
+  // not sure why reassignment is needed - its 2am and that makes it work... fix tomorrow!
+  Tree = Tree.addActions(({ flat }) => ({
+    updatePath(node, to) {
+      let from = node.path
+      node.path = to
+      if (from) delete flat[encode(from)]
+      flat[encode(to)] = node
+    },
+    reparentPaths(node, oldParent, newParent) {
+      treeUtils.walk(node => {
+        Tree.updatePath(node, [
+          ...newParent,
+          ...node.path.slice(oldParent.length),
+        ])
+      })(node)
+    },
+  }))
+  return {
+    getNode: Tree.getNode,
+    add: tree => Tree.add(tree.path, observable(blankNode())),
+    remove: (tree, node) => Tree.remove(node.path),
+    join: (tree, join) => Tree.mutate(tree.path, { join }),
+    mutate: Tree.mutate,
+    typeChange(node, type) {
+      let path = node.path
+      let parentPath = _.dropRight(1, node.path)
```

Comment:
This line has `node.path` but the previous line has `path = node.path`. Not sure what's better, if changing this specific line to use `path` or changing all the other lines to use `node.path` and removing `path = node.path` since it doesn't seem to add much value.